src/server.cpp

```diff
@@ -1603,49 +1609,74 @@ void Server::SendShowFormspecMessage(session_t peer_id, const std::string &forms
 	Send(&pkt);
 }
 
-// Spawns a particle on peer with peer_id
-void Server::SendSpawnParticle(session_t peer_id, u16 protocol_version,
-	const ParticleParameters &p)
+void Server::SendSpawnParticles(RemotePlayer *player,
+		const std::vector<ParticleParameters> &particles)
 {
 	static thread_local const float radius =
-			g_settings->getS16("max_block_send_distance") * MAP_BLOCKSIZE * BS;
+		g_settings->getS16("max_block_send_distance") * MAP_BLOCKSIZE * BS;
+	const float radius_sq = radius * radius;
 
-	if (peer_id == PEER_ID_INEXISTENT) {
-		std::vector<session_t> clients = m_clients.getClientIDs();
-		const v3f pos = p.pos * BS;
-		const float radius_sq = radius * radius;
-
-		for (const session_t client_id : clients) {
-			RemotePlayer *player = m_env->getPlayer(client_id);
-			if (!player)
-				continue;
-
-			PlayerSAO *sao = player->getPlayerSAO();
-			if (!sao)
-				continue;
+	PlayerSAO *sao = player->getPlayerSAO();
+	if (!sao)
+		return;
 
-			// Do not send to distant clients
-			if (sao->getBasePosition().getDistanceFromSQ(pos) > radius_sq)
-				continue;
+	RemoteClient *client = m_clients.getClientNoEx(player->getPeerId(), CS_Active);
+	if (!client)
+		return;
 
-			SendSpawnParticle(client_id, player->protocol_version, p);
+	std::ostringstream oss(std::ios::binary);
+	for (const auto &particle : particles) {
+		if (sao->getBasePosition().getDistanceFromSQ(particle.pos * BS) > radius_sq)
+			continue; // out of range
+
+		particle.serialize(oss, player->protocol_version);
+		if (player->protocol_version < 50) {
+			// Client only supports TOCLIENT_SPAWN_PARTICLE,
+			// so turn the written particle into a packet immediately
+			NetworkPacket pkt(TOCLIENT_SPAWN_PARTICLE, oss.tellp(), player->getPeerId());
+			pkt.putRawString(oss.str());
+			Send(&pkt);
+			oss.clear();
 		}
-		return;
 	}
-	assert(protocol_version != 0);
 
-	NetworkPacket pkt(TOCLIENT_SPAWN_PARTICLE, 0, peer_id);
+	if (oss.tellp() == 0)
+		return; // no batch to send
 
+	// Client supports TOCLIENT_SPAWN_PARTICLE_BATCH
+	assert(player->protocol_version >= 50);
+	std::string compressed_data;
 	{
-		// NetworkPacket and iostreams are incompatible...
-		std::ostringstream oss(std::ios_base::binary);
-		p.serialize(oss, protocol_version);
-		pkt.putRawString(oss.str());
+		std::ostringstream compressed_data_oss;
+		compressZstd(oss.str(), compressed_data_oss);
+		compressed_data = compressed_data_oss.str();
 	}
-
+	NetworkPacket pkt(TOCLIENT_SPAWN_PARTICLE_BATCH,
+			4 + compressed_data.size(), player->getPeerId());
+	pkt.putLongString(compressed_data);
```

Comment:
FWIW this compresses very well for the rather random input data produced by https://github.com/luanti-org/luanti/issues/16450#issuecomment-3238611438 :
```
before=15800, after=1638
before=15800, after=1646
before=15800, after=1650
before=15800, after=1656
before=15800, after=1641
```
:+1: